### PR TITLE
Adds put_stream, compression and breaks lib.rs into modules

### PR DESCRIPTION
**Pull request by andrebsguedes**
This PR implements streaming put requests, changes the streaming get API a little, adds optional compression/decompression for the streaming variants and breaks the huge `src/lib.rs` file into modules.

### CODE REVIEW

File: src/ops.rs [hunk 1]
```diff
@@ -0,0 +1,257 @@
+use crate::{CResult, Config, NotifyGuard, SQ, clients, dyn_connect, static_config, Request};
```
Comment (jfunstonRAI) [inline src/ops.rs]: Can we rename this file? Maybe `put_get.rs` or something?

File: src/stream.rs [hunk 2]
```diff
@@ -0,0 +1,584 @@
...
+pub struct ReadResponse {
+    result: CResult,
+    length: usize,
+    eof: bool,
+    error_message: *mut c_char
+}
+
+unsafe impl Send for ReadResponse {}
+
+// RAII Guard for a ListResponse that ensures the awaiting Julia task will be notified
```
Comment (jfunstonRAI) [inline src/stream.rs]: `ListResponse` typo?

File: src/stream.rs [hunk 3]
```diff
@@ -0,0 +1,584 @@
...
+            }
+        }
+        None => {
+            return CResult::Error;
+        }
+    }
+}
+
+#[no_mangle]
+pub extern "C" fn write_to_stream(
```
Comment (jfunstonRAI) [inline src/stream.rs:583]: I think we need some more documentation for these more complicated features, either as comments or in the README.
Does this handle chunking the puts into the size needed by each cloud provider?
Do you need to flush, or will shutdown do it for you?
How is concurrency handled (seems unbounded and needs to be controlled by the user)?

**Review by jfunstonRAI — CHANGES_REQUESTED** [review]

File: src/ops.rs (hunk 1 above)
Comment (andrebsguedes) [reply src/ops.rs]: What about `basic_ops.rs` (ops == operations) ? `put_get.rs` does not cover the delete op which is in the same file

File: src/stream.rs (hunk 2 above)
Comment (andrebsguedes) [reply src/stream.rs]: Will fix!

File: src/stream.rs (hunk 3 above)
Comment (andrebsguedes) [reply src/stream.rs:583]: I will document some of the more complex functions. But answering some of your questions:
- It splits the bytes written into 10 MB chunks which (according to their code) is allowed by the supported cloud providers.
- Only shutdown is required (it does flushing and some final completion request), flush is only exposed to give the user control over any internal buffering.
- It uploads up to 8 parts in parallel and the user has no control over this (it is hard-coded), also any interaction with the stream from the Julia side is not thread-safe, so no concurrency from a Julia perspective.

File: src/ops.rs (hunk 1 above)
Comment (jfunstonRAI) [reply src/ops.rs]: Maybe, it still seems a bit vague. How about `crud_ops.rs`? It doesn't perfectly match the cloud APIs but it gets the point across. Or move the delete into its own file even though it is small? Up to you

**Review by jfunstonRAI — APPROVED** [review]
Looks good!

File: src/ops.rs (hunk 1 above)
Comment (andrebsguedes) [reply src/ops.rs]: Changed it  to `crud_ops.rs`

**Pull request merged**